### PR TITLE
Introduce LSPInput abstraction to send input to multithreaded LSP.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Introduce LSPInput abstraction to send input to multithreaded LSP.

### Motivation

Will make it possible to use multithreaded LSP in other scenarios besides STDIN (e.g. LSPWrapper).

### Test plan

Manually tested Sorbet with this change. Also tested via existing recorded tests. Will be tested further once I modify LSPWrapper to support multithreaded mode.

### CODE REVIEW

File: main/lsp/LSPInput.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#include "spdlog/spdlog.h"
+#include <iterator>
+
+using namespace std;
+
+namespace sorbet::realmain::lsp {
+
+LSPFDInput::LSPFDInput(shared_ptr<spdlog::logger> logger, int inputFd) : logger(move(logger)), inputFd(inputFd) {}
+
+unique_ptr<LSPMessage> LSPFDInput::read(int timeoutMs) {
```
Comment (jvilk-stripe) [inline main/lsp/LSPInput.cc]: Moved from `getNewRequest` in `protocol.cc`

File: main/lsp/LSPInput.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    std::unique_ptr<LSPMessage> read(int timeoutMs) override;
+};
+
+/**
+ * Input is provided programmatically via the `write` method. Threadsafe.
+ *
+ * Throws a FileReadException when stream has been closed.
+ */
+class LSPProgrammaticInput final : public LSPInput {
```
Comment (jvilk-stripe) [inline main/lsp/LSPInput.h]: This class is unused so far. I plan to use it in LSPWrapper. Can delurk if it adds too much noise to PR.

File: main/lsp/lsp.h [hunk 3]
```diff
@@ -1,18 +1,11 @@
 #ifndef RUBY_TYPER_LSPLOOP_H
 #define RUBY_TYPER_LSPLOOP_H
 
-#include "ast/ast.h"
-#include "common/kvstore/KeyValueStore.h"
-#include "core/ErrorQueue.h"
-#include "core/NameHash.h"
```
Comment (jvilk-stripe) [inline main/lsp/lsp.h]: I noticed that `lsp.h` no longer needs to include many of these headers as I moved the relevant references to other classes, like `LSPTypechecker`.

File: main/realmain.cc [hunk 4]
```diff
@@ -164,15 +166,6 @@ core::Loc findTyped(unique_ptr<core::GlobalState> &gs, core::FileRef file) {
 }
 
 #ifndef SORBET_REALMAIN_MIN
-class LSPStdout final : public lsp::LSPOutput {
```
Comment (jvilk-stripe) [inline main/realmain.cc]: I'm a little embarrassed... in a previous PR I moved this class into `LSPOutput.{h,cc}`, but must've forgotten to remove it from realmain.

File: main/lsp/lsp.h [hunk 5]
```diff
@@ -25,6 +18,9 @@
 // This is an implementation of LSP protocol (version 3.13) for Sorbet
 namespace sorbet::realmain::lsp {
 
+class LSPInput;
```
Comment (DarkDimius) [inline main/lsp/lsp.h]: thank you for forward declaring them

File: main/lsp/LSPInput.cc [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    absl::MutexLock lock(&mtx);
+    if (closed) {
+        Exception::raise("Cannot write to a closed input.");
+    }
+    available.push_back(move(message));
+}
+
+void LSPProgrammaticInput::write(vector<unique_ptr<LSPMessage>> messages) {
+    absl::MutexLock lock(&mtx);
+    available.insert(available.end(), make_move_iterator(messages.begin()), make_move_iterator(messages.end()));
```
Comment (DarkDimius) [inline main/lsp/LSPInput.cc]: why doesn't this make `closed` check?

File: main/lsp/LSPInput.cc [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+void LSPProgrammaticInput::close() {
+    absl::MutexLock lock(&mtx);
+    if (closed) {
+        Exception::raise("Input already ended.");
```
Comment (DarkDimius) [inline main/lsp/LSPInput.cc]: thank you

File: main/lsp/LSPInput.cc [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    ENFORCE(buffer.length() >= length);
+
+    string json = buffer.substr(0, length);
+    buffer.erase(0, length);
+    logger->debug("Read: {}\n", json);
+    return LSPMessage::fromClient(json);
+}
+
+unique_ptr<LSPMessage> LSPProgrammaticInput::read(int timeoutMs) {
```
Comment (DarkDimius) [inline main/lsp/LSPInput.cc]: do we want timeout to be configurable?

File: main/lsp/LSPInput.cc [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    absl::MutexLock lock(&mtx);
+    if (closed && available.empty()) {
+        throw sorbet::FileReadException("EOF");
```
Comment (DarkDimius) [inline main/lsp/LSPInput.cc]: you expect this to be an exceptional case, right?

File: main/lsp/LSPInput.cc (hunk 8 above)
Comment (jvilk-stripe) [reply main/lsp/LSPInput.cc]: The alternative is encoding a magic constant in all subclasses. I think this is cleaner. There's a default value, too, in the abstract method.

File: main/lsp/LSPInput.cc (hunk 9 above)
Comment (jvilk-stripe) [reply main/lsp/LSPInput.cc]: Not really; this exception signals end-of-input. It matches the interface we had previously with `getNewRequest`.

Comment (jvilk-stripe) [reply main/lsp/LSPInput.cc]: I'm going to move forward with this design to unblock a subsequent PR -- I'm happy to change the interface in a future PR.

**Pull request merged**